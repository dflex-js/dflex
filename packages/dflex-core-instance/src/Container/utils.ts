import { getElmOverflow, PointBool, Axis } from "@dflex/utils";

const OVERFLOW_REGEX = /(auto|scroll|overlay)/;

type CSSOverflow = ReturnType<typeof getElmOverflow>;

function hasScrollableContent(axis: Axis, DOM: HTMLElement): boolean {
  return axis === "y"
    ? DOM.scrollHeight > DOM.clientHeight
    : DOM.scrollWidth > DOM.clientWidth;
}

function hasOverflowProperty(axis: Axis, DOM: HTMLElement): boolean {
  const overflow: CSSOverflow = getElmOverflow(
    DOM,
    axis === "x" ? "overflow-x" : "overflow-y",
  );

  return OVERFLOW_REGEX.test(overflow);
}

function hasOverflow(axis: Axis, DOM: HTMLElement): boolean {
  if (hasOverflowProperty(axis, DOM)) {
    return hasScrollableContent(axis, DOM);
  }

  return false;
}
