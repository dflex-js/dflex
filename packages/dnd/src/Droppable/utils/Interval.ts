/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

class Interval {
  intervalID: number | null;

  constructor() {
    this.intervalID = 0;
  }

  set(cb: Function, delay: number) {
    this.intervalID = setInterval(cb, delay);
  }

  clear() {
    clearInterval(this.intervalID!);
    this.intervalID = null;
  }
}

export default Interval;
