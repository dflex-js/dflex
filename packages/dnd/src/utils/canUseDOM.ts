/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

function canUseDOM() {
  return (
    typeof window !== "undefined" &&
    typeof window.document !== "undefined" &&
    typeof window.document.createElement !== "undefined"
  );
}

export default canUseDOM;
