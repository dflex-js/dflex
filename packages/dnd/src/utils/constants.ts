import type { DefaultDndOpts } from "../types";

export function noop() {}

export const defaultOpts: DefaultDndOpts = Object.freeze({
  enableContainersTransition: true,

  threshold: {
    vertical: 60,
    horizontal: 60,
  },

  restrictions: {
    self: {
      allowLeavingFromTop: true,
      allowLeavingFromBottom: true,
      allowLeavingFromLeft: true,
      allowLeavingFromRight: true,
    },
    container: {
      allowLeavingFromTop: true,
      allowLeavingFromBottom: true,
      allowLeavingFromLeft: true,
      allowLeavingFromRight: true,
    },
  },

  scroll: {
    enable: true,
    initialSpeed: 10,
    threshold: {
      vertical: 15,
      horizontal: 15,
    },
  },

  events: {
    onDragOver: noop,
    onDragLeave: noop,
    onDragOutContainer: noop,
    onDragOutThreshold: noop,
    onLiftUpSiblings: noop,
    onMoveDownSiblings: noop,
    onStateChange: noop,
  },
});
