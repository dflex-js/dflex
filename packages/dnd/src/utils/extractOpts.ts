/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

// @ts-nocheck

// TODO: Do you love typescript? Here we go buddy. It's your chance. Start your
// PR. Don't be shy. I'll be waiting.

import type { DndOpts, FinalDndOpts } from "../types";

export const defaultOpts: DndOpts = Object.freeze({
  threshold: {
    vertical: 60,
    horizontal: 60,
  },

  restrictions: {
    self: {
      allowLeavingFromTop: true,
      allowLeavingFromBottom: true,
      allowLeavingFromLeft: true,
      allowLeavingFromRight: true,
    },
    container: {
      allowLeavingFromTop: true,
      allowLeavingFromBottom: true,
      allowLeavingFromLeft: true,
      allowLeavingFromRight: true,
    },
  },

  scroll: {
    enable: true,
    initialSpeed: 10,
    threshold: {
      vertical: 75,
      horizontal: 75,
    },
  },
});

export function extractOpts(opts: DndOpts) {
  const options = { ...opts };

  (Object.keys(defaultOpts) as Array<keyof FinalDndOpts>).forEach((props) => {
    if (props === "restrictions") {
      options.restrictionsStatus = {
        isSelfRestricted: false,
        isContainerRestricted: false,
      };
    }

    if (!options[props]) {
      options[props] = { ...defaultOpts[props] };

      return;
    }

    Object.keys(defaultOpts[props]).forEach((subProp) => {
      if (typeof defaultOpts[props][subProp] === "object") {
        if (options[props][subProp] !== undefined) {
          options[props][subProp] = {
            ...defaultOpts[props][subProp],
            ...options[props][subProp],
          };

          return;
        }

        options[props][subProp] = { ...defaultOpts[props][subProp] };

        return;
      }

      if (options[props][subProp] === undefined) {
        options[props][subProp] = defaultOpts[props][subProp];
      }
    });

    if (props === "restrictions") {
      const restrictionsKeys = Object.keys(options.restrictions.self);

      let isSelfRestricted = false;

      for (let i = 0; i < restrictionsKeys.length; i += 1) {
        if (!options[props].self[restrictionsKeys[i]]) {
          isSelfRestricted = true;

          break;
        }
      }

      let isContainerRestricted = false;

      for (let i = 0; i < restrictionsKeys.length; i += 1) {
        if (!options[props].container[restrictionsKeys[i]]) {
          isContainerRestricted = true;

          break;
        }
      }

      options.restrictionsStatus.isSelfRestricted = isSelfRestricted;
      options.restrictionsStatus.isContainerRestricted = isContainerRestricted;
    }
  });

  return options as FinalDndOpts;
}

export default extractOpts;
