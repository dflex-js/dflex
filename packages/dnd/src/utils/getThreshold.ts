/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type { ThresholdPercentages, ThresholdInPixels } from "../types";

function getThreshold(
  thresholdInput: ThresholdPercentages,
  containerWidth: number,
  containerHeight: number
): ThresholdInPixels {
  const x = Math.round((thresholdInput.horizontal * containerWidth) / 100);

  const y = Math.round((thresholdInput.vertical * containerHeight) / 100);

  return {
    y,
    x,
  };
}

export default getThreshold;
