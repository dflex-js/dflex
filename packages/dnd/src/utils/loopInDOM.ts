function loopInDOM(
  fromElement: HTMLElement,
  // eslint-disable-next-line no-unused-vars
  cb: (arg: HTMLElement) => boolean
) {
  let current: HTMLElement | null = fromElement;

  do {
    if (cb(current)) {
      return current;
    }

    current = current.parentNode as HTMLElement;
  } while (current && !current.isSameNode(document.body));

  return null;
}

export default loopInDOM;
