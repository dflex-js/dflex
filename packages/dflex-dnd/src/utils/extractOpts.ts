// @ts-nocheck

// TODO: Add types for this.

import type { DndOpts, FinalDndOpts } from "../types";
import { defaultOpts, noop } from "./constants";

export function extractOpts(opts: DndOpts) {
  const options = { ...opts };

  (Object.keys(defaultOpts) as Array<keyof FinalDndOpts>).forEach((props) => {
    if (props === "restrictions") {
      options.restrictionsStatus = {
        isSelfRestricted: false,
        isContainerRestricted: false,
      };
    }

    if (options[props] === undefined) {
      options[props] =
        typeof options[props] === "object"
          ? { ...defaultOpts[props] }
          : defaultOpts[props];

      return;
    }

    Object.keys(defaultOpts[props]).forEach((subProp) => {
      // for sub options with objects values, like restrictions.
      if (typeof defaultOpts[props][subProp] === "object") {
        if (options[props][subProp] !== undefined) {
          options[props][subProp] = {
            ...defaultOpts[props][subProp],
            ...options[props][subProp],
          };

          return;
        }

        options[props][subProp] = { ...defaultOpts[props][subProp] };

        return;
      }

      if (options[props][subProp] === undefined) {
        options[props][subProp] = defaultOpts[props][subProp];
      }
    });

    if (props === "restrictions") {
      const restrictionsKeys = Object.keys(options.restrictions.self);

      let isSelfRestricted = false;

      for (let i = 0; i < restrictionsKeys.length; i += 1) {
        if (!options[props].self[restrictionsKeys[i]]) {
          isSelfRestricted = true;

          break;
        }
      }

      let isContainerRestricted = false;

      for (let i = 0; i < restrictionsKeys.length; i += 1) {
        if (!options[props].container[restrictionsKeys[i]]) {
          isContainerRestricted = true;

          break;
        }
      }

      options.restrictionsStatus.isSelfRestricted = isSelfRestricted;
      options.restrictionsStatus.isContainerRestricted = isContainerRestricted;
    } else if (props === "events") {
      Object.keys(options.events).forEach((event) => {
        if (typeof options.events[event] !== "function") {
          options.events[event] = noop;
        }
      });
    }
  });

  return options as FinalDndOpts;
}

export default extractOpts;
