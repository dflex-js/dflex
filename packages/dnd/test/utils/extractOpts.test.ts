/**
 * Copyright (c) Jalal Maskoun.
 *
 * This source code is licensed under the AGPL3.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { extractOpts } from "../../src/utils/extractOpts";

describe("extractOpts", () => {
  it("Returns default options when empty object is passed", () => {
    const opts = extractOpts({});

    expect(opts).toMatchSnapshot();
  });

  it("Returns the default options with input scroll/threshold", () => {
    const opts = extractOpts({ scroll: { threshold: { horizontal: 1000 } } });

    expect(opts).toMatchSnapshot();

    expect(opts.scroll.threshold.horizontal).toEqual(1000);
  });

  it("Returns the default options with input scroll/threshold/enable", () => {
    const opts = extractOpts({
      scroll: { enable: false, threshold: { horizontal: 1000 } },
    });

    expect(opts).toMatchSnapshot();

    expect(opts.scroll.threshold.horizontal).toEqual(1000);
    expect(opts.scroll.enable).toEqual(false);
  });

  it("Returns the default options with input restrictions/allowLeavingFromBottom", () => {
    const opts = extractOpts({
      restrictions: {
        allowLeavingFromBottom: false,
      },
    });

    expect(opts).toMatchSnapshot();

    expect(opts.restrictions.allowLeavingFromBottom).toEqual(false);
  });
});

it("Returns the default options with input restrictions/threshold/scroll", () => {
  const opts = extractOpts({
    restrictions: {
      allowLeavingFromBottom: false,
    },
    threshold: {
      horizontal: 9000,
    },
    scroll: {
      threshold: {
        horizontal: 9000,
      },
    },
  });

  expect(opts).toMatchSnapshot();

  expect(opts.restrictions.allowLeavingFromBottom).toEqual(false);
  expect(opts.threshold.horizontal).toEqual(9000);
  expect(opts.scroll.threshold.horizontal).toEqual(9000);
});
