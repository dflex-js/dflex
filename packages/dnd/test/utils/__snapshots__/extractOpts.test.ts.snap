// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Returns the default options with input restrictions/threshold/scroll 1`] = `
Object {
  "restrictions": Object {
    "allowLeavingFromBottom": false,
    "allowLeavingFromLeft": true,
    "allowLeavingFromRight": true,
    "allowLeavingFromTop": true,
  },
  "scroll": Object {
    "enable": true,
    "speed": 10,
    "threshold": Object {
      "horizontal": 9000,
      "vertical": 80,
    },
  },
  "threshold": Object {
    "horizontal": 9000,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns default options when empty object is passed 1`] = `
Object {
  "restrictions": Object {
    "allowLeavingFromBottom": true,
    "allowLeavingFromLeft": true,
    "allowLeavingFromRight": true,
    "allowLeavingFromTop": true,
  },
  "scroll": Object {
    "enable": true,
    "speed": 10,
    "threshold": Object {
      "horizontal": 80,
      "vertical": 80,
    },
  },
  "threshold": Object {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input restrictions/allowLeavingFromBottom 1`] = `
Object {
  "restrictions": Object {
    "allowLeavingFromBottom": false,
    "allowLeavingFromLeft": true,
    "allowLeavingFromRight": true,
    "allowLeavingFromTop": true,
  },
  "scroll": Object {
    "enable": true,
    "speed": 10,
    "threshold": Object {
      "horizontal": 80,
      "vertical": 80,
    },
  },
  "threshold": Object {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input scroll/threshold 1`] = `
Object {
  "restrictions": Object {
    "allowLeavingFromBottom": true,
    "allowLeavingFromLeft": true,
    "allowLeavingFromRight": true,
    "allowLeavingFromTop": true,
  },
  "scroll": Object {
    "enable": true,
    "speed": 10,
    "threshold": Object {
      "horizontal": 1000,
      "vertical": 80,
    },
  },
  "threshold": Object {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;

exports[`extractOpts Returns the default options with input scroll/threshold/enable 1`] = `
Object {
  "restrictions": Object {
    "allowLeavingFromBottom": true,
    "allowLeavingFromLeft": true,
    "allowLeavingFromRight": true,
    "allowLeavingFromTop": true,
  },
  "scroll": Object {
    "enable": false,
    "speed": 10,
    "threshold": Object {
      "horizontal": 1000,
      "vertical": 80,
    },
  },
  "threshold": Object {
    "horizontal": 60,
    "vertical": 60,
  },
}
`;
